enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -3976,6 +3979,150 @@ class AdjointGenerator
+              PointerType::getUnqual(innerType), 0, 0, call.getArgOperand(4));
+          auto malins = CallInst::CreateMalloc(
+              gutils->getNewFromOriginal(&call), size->getType(), innerType,
+              size, call.getArgOperand(0), nullptr, "");
+          arg2 =
+              BuilderZ.CreateBitCast(malins, call.getArgOperand(3)->getType());
+          SmallVector<Value *, 3> args;
+          args.push_back(arg2);
+          args.push_back(gutils->getNewFromOriginal(call.getArgOperand(3)));
+          args.push_back(call.getArgOperand(0));
+          BuilderZ.CreateCall(dmemcpy, args);
+        }
+        if (xcache && ycache) {
+          auto valins1 = BuilderZ.CreateInsertValue(undefinit, arg1, 0);
+          cacheval = BuilderZ.CreateInsertValue(valins1, arg2, 1);
+        } else if (xcache)
+          cacheval = arg1;
+        else if (ycache)
+          cacheval = arg2;
+        gutils->cacheForReverse(BuilderZ, cacheval,
+                                getIndex(&call, CacheType::Tape));
+      }
+      if (Mode == DerivativeMode::ReverseModeCombined ||
+          Mode == DerivativeMode::ReverseModeGradient) {
+        IRBuilder<> Builder2(call.getParent());
+        getReverseBuilder(Builder2);
+        auto derivcall = gutils->oldFunc->getParent()->getOrInsertFunction(
+            dfuncName, Builder2.getVoidTy(), Builder2.getInt32Ty(), innerType,
+            call.getArgOperand(1)->getType(), Builder2.getInt32Ty(),
+            call.getArgOperand(3)->getType(), Builder2.getInt32Ty());
+        Value *structarg1 = lookup(
+            gutils->getNewFromOriginal(orig->getArgOperand(1)), Builder2);
+        Value *structarg2 = lookup(
+            gutils->getNewFromOriginal(orig->getArgOperand(3)), Builder2);
+        if (xcache || ycache) {
+          if (Mode == DerivativeMode::ReverseModeGradient &&
+              (!gutils->isConstantValue(call.getArgOperand(1)) ||
+               !gutils->isConstantValue(call.getArgOperand(3)))) {
+            cacheval = Builder2.CreatePHI(cachetype, 0);
+          }
+          cacheval = gutils->cacheForReverse(Builder2, cacheval,
+                                             getIndex(&call, CacheType::Tape));
+          if (xcache && ycache) {
+            structarg1 = BuilderZ.CreateExtractValue(cacheval, 0);
+            structarg2 = BuilderZ.CreateExtractValue(cacheval, 1);
+          } else if (xcache)
+            structarg1 = cacheval;
+          else if (ycache)
+            structarg2 = cacheval;
+        }
+        CallInst *firstdcall, *seconddcall;
+        if (!gutils->isConstantValue(call.getArgOperand(3))) {
+          assert(!isa<Constant>(orig->getArgOperand(3)));
+          SmallVector<Value *, 6> args1 = {
+              lookup(gutils->getNewFromOriginal(orig->getArgOperand(0)),
+                     Builder2),
+              diffe(orig, Builder2),
+              structarg1,
+              Builder2.getInt32(1),
```

Comment:
This should be the actual stride of x, unless it's cached at which point it can be 1.